core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
+///
+/// Conceptually, a [`FileServer`] is a sequence of `Rewriter`s, which transform
+/// a path from a request to a final response. [`FileServer`] add a set of default
+/// `Rewriter`s, which filter out dotfiles, apply a root path, normalize directories,
+/// and use `index.html`.
+///
+/// After running the chain of `Rewriter`s,
+/// [`FileServer`] uses the final [`Option<FileResponse>`](FileResponse)
+/// to respond to the request. If the response is `None`, a path that doesn't
+/// exist or a directory path, [`FileServer`] will respond with a
+/// [`Status::NotFound`](crate::http::Status::NotFound). Otherwise the [`FileServer`]
+/// will respond with a redirect or the contents of the file specified.
+///
+/// [`FileServer`] provides several helper methods to add `Rewriter`s:
+/// - [`FileServer::and_rewrite()`]
+/// - [`FileServer::filter_file()`]
+/// - [`FileServer::map_file()`]
+pub trait Rewriter: Send + Sync + 'static {
+    /// Alter the [`FileResponse`] as needed.
+    fn rewrite<'p, 'h>(&self, path: Option<FileResponse<'p, 'h>>) -> Option<FileResponse<'p, 'h>>;
+}
+
+/// A Response from a [`FileServer`]
+#[derive(Debug)]
+pub enum FileResponse<'p, 'h> {
+    /// Return the contents of the specified file.
+    File(File<'p, 'h>),
+    /// Returns a Redirect to the specified path. This needs to be an absolute
+    /// URI, so you should start with [`File.full_uri`](File) when constructing
+    /// a redirect.
+    Redirect(Redirect),
+}
 
-        FileServer { root: path.into(), options, rank: Self::DEFAULT_RANK }
+impl<'p, 'h> From<File<'p, 'h>> for FileResponse<'p, 'h> {
+    fn from(value: File<'p, 'h>) -> Self {
+        Self::File(value)
     }
+}
+impl<'p, 'h> From<File<'p, 'h>> for Option<FileResponse<'p, 'h>> {
+    fn from(value: File<'p, 'h>) -> Self {
+        Some(FileResponse::File(value))
+    }
+}
 
-    /// Sets the rank for generated routes to `rank`.
-    ///
-    /// # Example
-    ///
-    /// ```rust,no_run
-    /// use rocket::fs::{FileServer, Options};
-    ///
-    /// // A `FileServer` created with `from()` with routes of rank `3`.
-    /// FileServer::from("/public").rank(3);
-    ///
-    /// // A `FileServer` created with `new()` with routes of rank `-15`.
-    /// FileServer::new("/public", Options::Index).rank(-15);
-    /// ```
-    pub fn rank(mut self, rank: isize) -> Self {
-        self.rank = rank;
+impl<'p, 'h> From<Redirect> for FileResponse<'p, 'h> {
+    fn from(value: Redirect) -> Self {
+        Self::Redirect(value)
+    }
+}
+impl<'p, 'h> From<Redirect> for Option<FileResponse<'p, 'h>> {
+    fn from(value: Redirect) -> Self {
+        Some(FileResponse::Redirect(value))
+    }
+}
+
+/// A File response from a [`FileServer`]
+#[derive(Debug)]
+pub struct File<'p, 'h> {
+    /// The path to the file that [`FileServer`] will respond with.
+    pub path: Cow<'p, Path>,
+    /// The original uri from the [`Request`]
+    pub full_uri: &'p Origin<'p>,
```

Comment:
What is the utility of having `full_uri` here? I think we should be able to remove this, and places where this is being can likely be improved.